_posts/2018-10-11-migrating-from-galen-to-backstop.md

```diff
@@ -0,0 +1,229 @@
+           });
+   
+           await Promise.all(promisesForLoadingImages);
+       })()`);
+   };
+   ```
+
+It proved to be the least flaky out of the three.
+
+#### Approach no. 3.1 - add the timeout, yet again
+
+As it turned out, the CPU usage was still an issue, so we had to add this timeout. Luckily, not that much this time (100ms).
+
+#### Final solution
+
+Eventually, this was our solution for lazy loaded images:
+ 
+1. Scroll down to the bottom of the page; do it in steps (viewport height), with a small delay, 50ms, between the scrolls, so that the CPU has a time frame to alter the DOM if necessary.
+1. Scroll back up, to the top of the page - by that time all of the lazy loaded elements were triggered and DOM was altered.
+1. Attach event listeners (`load` event) to the `img` elements.
+1. Wait for all of the events to resolve.
+1. Wait a bit more (the 100ms time).
+
+Done.
+
+## Switch
+
+For some time we were running both (galen and backstop based) visual regression tests in parallel. After having the backstop one stabilized, we turned the other one off.
+
+## Git repository and storing reference files
+
+One of the things we perceive as crucial is the possibility to clone the repo of our application and with just a few commands run it or execute the tests. It should be self contained.
+
+What about storing reference files inside the repository, then? Should we keep them there? We decided **not to**, and to put them on the server, where all of the tests are executed. Why? For two key reasons.
+
+### 1. Differences in rendering
+
+Despite the fact, that all of the tests are run in the same browser everywhere (headless Chrome), it is not guaranteed that the rendering will look 100% the same on various OSs. E.g. I am using a macOS, my colleague is using some Linux distro. I bet there will be differences. If he were to submit new reference files, tests on my end would fail. So could the ones executed on the pipeline server (yet another OS).
+
+Therefore we decided to keep the files on the pipeline server, where the whole environment is constant.
+
+### 2. Large amount of files
+
+Right now, our reference files weight ~85MB. Let's say something gets changed in the header of the application. Most of the images would have to be refreshed and, say, ~60MB of new ones would be created. What would that mean? Downloading that ~60MB of data on next `git pull`, or whooping ~145MB of data on `git clone`. What if more changes are applied, or reverted? More and more MBs to download. That would be just too much.
+
+Sure, there are ways of bypassing this. One of the solutions is to use [LFS](https://git-lfs.github.com/). Yet, using that extension doesn't fix the no. 1 problem (rendering). We could go a bit further and use [Docker](https://www.docker.com/) to ensure we have the very same environment everywhere, but that just adds too much complexity 'only' for running visual regression tests.
+
```

Comment:
There's also one problem, which isn't solved yet but it's worth mentioning. Usually, visual regression tests have threshold - amount of pixels which might differ. Sometimes intentional change isn't detected (for example, changing size of the X mark in close button) for big images (like 3-page-long desktop page). 

It's not a big issue, just something which I believe is worth mentioning and have to be kept in mind.

Reply:
You're right. More details === more help to the reader(s) === better job done. Thanks.